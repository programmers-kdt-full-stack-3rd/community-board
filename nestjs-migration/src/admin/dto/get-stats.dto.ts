import { Type } from "class-transformer";
import { IsDateString, IsOptional, IsString, Matches } from "class-validator";
import { TInterval } from "shared";

export class GetStatsQueryDto {
	@IsOptional()
	@IsDateString({}, { message: "startDate가 잘못되었습니다." })
	@Matches(/^\d{4}-\d{2}-\d{2}$/, { message: "startDate가 잘못되었습니다." })
	@Type(() => String)
	startDate?: string;

	@IsOptional()
	@IsDateString({}, { message: "endDate가 잘못되었습니다." })
	@Matches(/^\d{4}-\d{2}-\d{2}$/, { message: "endDate가 잘못되었습니다." })
	@Type(() => String)
	endDate?: string;

	@IsOptional()
	@IsString()
	interval?: TInterval = "daily";
}
