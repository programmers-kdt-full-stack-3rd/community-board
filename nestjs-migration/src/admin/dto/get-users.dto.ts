import { Transform } from "class-transformer";
import { IsOptional } from "class-validator";

export class GetUsersDto {
	@IsOptional()
	@Transform(({ value }) => parseInt(value, 10))
	index?: number = 0;

	@IsOptional()
	@Transform(({ value }) => parseInt(value, 10))
	perPage?: number = 10;

	@IsOptional()
	email?: string;

	@IsOptional()
	nickname?: string;
}
