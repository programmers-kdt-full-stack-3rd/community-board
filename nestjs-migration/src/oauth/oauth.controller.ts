import {
	Body,
	Controller,
	Delete,
	Get,
	HttpCode,
	HttpStatus,
	Param,
	Post,
	Res,
	UseGuards,
} from "@nestjs/common";
import { Response } from "express";
import { User } from "../common/decorator/user.decorator";
import { LoginGuard } from "../common/guard/login.guard";
import { IUserEntity } from "../common/interface/user-entity.interface";
import { COOKIE_MAX_AGE } from "../user/constant/user.constants";
import { getKstNow } from "../utils/date.util";
import { OAuthLoginDto } from "./dto/oauth-login.dto";
import { OAuthProviderDto } from "./dto/provider-param.dto";
import { OAuthService } from "./oauth.service";

@Controller("oauth")
export class OAuthController {
	constructor(private readonly oauthService: OAuthService) {}

	@Get("/login-url/:provider")
	@HttpCode(HttpStatus.OK)
	getLoginUrl(@Param() param: OAuthProviderDto) {
		const provider = param.provider;
		const url = this.oauthService.getOAuthUrl("login", provider);

		return { url };
	}

	@Post("/login")
	@HttpCode(HttpStatus.OK)
	async oAuthlogin(
		@Body() oAuthLoginDto: OAuthLoginDto,
		@Res({ passthrough: true }) res: Response
	) {
		const result = await this.oauthService.oAuthLogin(oAuthLoginDto);

		res.cookie("accessToken", result.accessToken, {
			httpOnly: true,
			secure: true,
			maxAge: COOKIE_MAX_AGE.accessToken,
		});

		res.cookie("refreshToken", result.refreshToken, {
			httpOnly: true,
			secure: true,
			maxAge: COOKIE_MAX_AGE.refreshToken,
		});

		return {
			message: "로그인 성공",
			result: { nickname: result.nickname, loginTime: getKstNow() },
		};
	}

	@Get("/reconfirm-url/:provider")
	@HttpCode(HttpStatus.OK)
	getReconfirmUrl(@Param() param: OAuthProviderDto) {
		const provider = param.provider;
		const url = this.oauthService.getOAuthUrl("reconfirm", provider);

		return { url };
	}

	@Post("/reconfirm")
	@HttpCode(HttpStatus.OK)
	@UseGuards(LoginGuard)
	async oAuthReconfirm(
		@Body() oAuthLoginDto: OAuthLoginDto,
		@Res({ passthrough: true }) res: Response,
		@User() user: IUserEntity
	) {
		const { tempToken } = await this.oauthService.oAuthReconfirm(
			oAuthLoginDto,
			user.userId
		);

		res.cookie("tempToken", tempToken, {
			httpOnly: true,
			secure: true,
			maxAge: COOKIE_MAX_AGE.tempToken,
		});

		return { message: "재확인 성공" };
	}

	@Get("/link-url/:provider")
	@HttpCode(HttpStatus.OK)
	getLinkUrl(@Param() param: OAuthProviderDto) {
		const provider = param.provider;
		const url = this.oauthService.getOAuthUrl("link", provider);

		return { url };
	}

	@Post("/link")
	@HttpCode(HttpStatus.OK)
	@UseGuards(LoginGuard)
	async oAuthLink(
		@Body() oAuthLoginDto: OAuthLoginDto,
		@User() user: IUserEntity
	) {
		await this.oauthService.oAuthLink(oAuthLoginDto, user.userId);

		return { message: "소셜 계정 연동 성공" };
	}

	@Delete("/link/:provider")
	@HttpCode(HttpStatus.OK)
	@UseGuards(LoginGuard)
	async oAuthUnlink(
		@Param() param: OAuthProviderDto,
		@User() user: IUserEntity
	) {
		const provider = param.provider;

		await this.oauthService.oAuthUnlink(provider, user.userId);

		return { message: "소셜 계정 연동 해제 성공" };
	}
}
