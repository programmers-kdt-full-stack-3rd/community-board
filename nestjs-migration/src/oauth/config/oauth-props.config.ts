import { Injectable } from "@nestjs/common";
import { ConfigService } from "@nestjs/config";
import { TOAuthProvider } from "shared";
import { oAuthRequestContentType } from "../constants/oauth.constants";
import { TOAuthProps } from "../interfaces/oauth.interface";

@Injectable()
export class OAuthPropsConfig {
	constructor(private readonly configService: ConfigService) {}

	getOAuthProps(): Record<TOAuthProvider, TOAuthProps> {
		const getRedirectUri = (provider: TOAuthProvider) => {
			const port = this.configService.get<string>("app.port");
			const baseUrl = `http://localhost:${port}/oauth/redirect`;

			return `${baseUrl}/${provider}`;
		};

		const googleClientId = this.configService.get<string>(
			"oauth.google.client_id"
		);
		const googleClientSecret = this.configService.get<string>(
			"oauth.google.client_secret"
		);

		const kakaoClientId = this.configService.get<string>(
			"oauth.kakao.client_id"
		);
		const kakaoClientSecret = this.configService.get<string>(
			"oauth.kakao.client_secret"
		);

		const naverClientId = this.configService.get<string>(
			"oauth.naver.client_id"
		);
		const naverClientSecret = this.configService.get<string>(
			"oauth.naver.client_secret"
		);

		return {
			google: {
				clientId: googleClientId,
				clientSecret: googleClientSecret,

				scope: "https://www.googleapis.com/auth/userinfo.profile",
				redirectUri: getRedirectUri("google"),

				requestEndpoint: {
					login: "https://accounts.google.com/o/oauth2/v2/auth",
					token: "https://oauth2.googleapis.com/token",
					user: "https://www.googleapis.com/oauth2/v2/userinfo",
					revoke: "https://oauth2.googleapis.com/revoke",
				},

				getAdditionalRequestOptionsFor: {
					login: () => ({
						searchParams: {
							access_type: "offline",
						},
					}),

					revoke: ({ accessToken }: { accessToken: string }) => ({
						searchParams: {
							token: accessToken,
						},
					}),
				},

				reconfirmParams: {
					prompt: "consent",
				},
			},

			kakao: {
				clientId: kakaoClientId,
				clientSecret: kakaoClientSecret,

				redirectUri: getRedirectUri("kakao"),

				requestEndpoint: {
					login: "https://kauth.kakao.com/oauth/authorize",
					token: "https://kauth.kakao.com/oauth/token",
					user: "https://kapi.kakao.com/v2/user/me",
					revoke: "https://kapi.kakao.com/v1/user/unlink",
				},

				getAdditionalRequestOptionsFor: {
					revoke: ({ accessToken }: { accessToken: string }) => ({
						headers: {
							Authorization: `Bearer ${accessToken}`,
						},
					}),
				},

				reconfirmParams: {
					prompt: "login",
				},
			},

			naver: {
				clientId: naverClientId,
				clientSecret: naverClientSecret,

				redirectUri: getRedirectUri("naver"),

				requestEndpoint: {
					login: "https://nid.naver.com/oauth2.0/authorize",
					token: "https://nid.naver.com/oauth2.0/token",
					user: "https://openapi.naver.com/v1/nid/me",
					revoke: "https://nid.naver.com/oauth2.0/token",
				},

				getAdditionalRequestOptionsFor: {
					revoke: ({ accessToken }: { accessToken: string }) => ({
						headers: {
							"Content-type": oAuthRequestContentType,
						},
						body: {
							grant_type: "delete",
							client_id: naverClientId,
							client_secret: naverClientSecret,
							access_token: accessToken,
						},
					}),
				},

				reconfirmParams: {
					auth_type: "reauthenticate",
				},
			},
		};
	}
}
