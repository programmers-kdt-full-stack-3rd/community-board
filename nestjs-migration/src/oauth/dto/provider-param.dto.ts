import { IsIn } from "class-validator";
import { TOAuthProvider, oAuthProviders } from "shared";
import { VALIDATION_ERROR_MESSAGES } from "../constants/oauth.constants";

export class OAuthProviderDto {
	@IsIn(oAuthProviders, {
		message: VALIDATION_ERROR_MESSAGES.INVALID_OAUTH_PROVIDER,
	})
	provider: TOAuthProvider;
}
