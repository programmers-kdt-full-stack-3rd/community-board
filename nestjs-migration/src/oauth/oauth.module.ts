import { Module } from "@nestjs/common";
import { AuthModule } from "../auth/auth.module";
import { UserRepository } from "../user/user.repository";
import { OAuthPropsConfig } from "./config/oauth-props.config";
import { OAuthController } from "./oauth.controller";
import { OAuthService } from "./oauth.service";
import { OAuthConnectionRepository } from "./repositories/oauth-connection.repository";
import { OAuthProviderRepository } from "./repositories/oauth-provider.repository";

@Module({
	imports: [AuthModule],
	controllers: [OAuthController],
	providers: [
		OAuthService,
		OAuthPropsConfig,
		OAuthConnectionRepository,
		OAuthProviderRepository,
		UserRepository,
	],
})
export class OAuthModule {}
