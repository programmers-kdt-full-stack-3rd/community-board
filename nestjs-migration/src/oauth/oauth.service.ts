import { Injectable } from "@nestjs/common";
import { stringify } from "node:querystring";
import { TOAuthProvider } from "shared";
import { OAuthPropsConfig } from "./config/oauth-props.config";
import { TOAuthLoginType } from "./interfaces/oauth.interface";

@Injectable()
export class OAuthService {
	constructor(private oAuthPropsConfig: OAuthPropsConfig) {}

	getOAuthUrl(loginType: TOAuthLoginType, provider: TOAuthProvider) {
		const { [loginType]: url } = this.buildLoginUrl(provider);

		return url;
	}

	private buildOAuthState = (loginType: TOAuthLoginType) => {
		return stringify({
			login_type: loginType,
		});
	};

	private buildLoginUrl(provider: TOAuthProvider): {
		[key in TOAuthLoginType]: string;
	} {
		const {
			requestEndpoint,
			clientId,
			redirectUri,
			scope,
			getAdditionalRequestOptionsFor,
			reconfirmParams,
		} = this.oAuthPropsConfig.getOAuthProps()[provider];

		const loginUrl = new URL(requestEndpoint.login);

		loginUrl.searchParams.set("response_type", "code");
		loginUrl.searchParams.set("client_id", clientId);
		loginUrl.searchParams.set("redirect_uri", redirectUri);
		loginUrl.searchParams.set("state", this.buildOAuthState("login"));

		if (scope) {
			loginUrl.searchParams.set("scope", scope);
		}

		if (getAdditionalRequestOptionsFor?.login) {
			const { searchParams = {} } =
				getAdditionalRequestOptionsFor.login();

			for (const key in searchParams) {
				loginUrl.searchParams.set(key, searchParams[key]);
			}
		}

		const reconfirmUrl = new URL(loginUrl);
		reconfirmUrl.searchParams.set(
			"state",
			this.buildOAuthState("reconfirm")
		);
		for (const key in reconfirmParams) {
			reconfirmUrl.searchParams.set(key, reconfirmParams[key]);
		}

		const linkUrl = new URL(loginUrl);
		linkUrl.searchParams.set("state", this.buildOAuthState("link"));

		return {
			login: loginUrl.toString(),
			reconfirm: reconfirmUrl.toString(),
			link: linkUrl.toString(),
		};
	}
}
