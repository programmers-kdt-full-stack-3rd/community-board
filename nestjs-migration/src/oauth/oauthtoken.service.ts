import { Injectable } from "@nestjs/common";
import { TOAuthProvider } from "shared";
import { ServerError } from "../common/exceptions/server-error.exception";
import {
	buildTokenFetchOptions,
	extractOAuthAccountId,
} from "../utils/oauth.util";
import { OAuthPropsConfig } from "./config/oauth-props.config";
import { OAUTH_TOKEN_SERVICE_ERROR_MESSAGES } from "./constants/oauth-token-service.constatns";
import { oAuthRequestContentType } from "./constants/oauth.constants";
import {
	IOAuthTokens,
	TOAuthTokenRequestGrantType,
} from "./interfaces/oauth.interface";

@Injectable()
export class OAuthTokenService {
	constructor(private oAuthPropsConfig: OAuthPropsConfig) {}

	private get oAuthProps() {
		return this.oAuthPropsConfig.getOAuthProps();
	}

	async verifyAuthorizationCode(
		provider: TOAuthProvider,
		authorizationCode: string
	) {
		const oAuthTokens = await this.fetchOAuthTokens(
			provider,
			"authorization_code",
			authorizationCode
		);

		const oAuthUser = await this.fetchOAuthUserByAccessToken(
			provider,
			oAuthTokens.access_token
		);

		return {
			oAuthAccountId: extractOAuthAccountId(provider, oAuthUser),
			oAuthRefreshToken: oAuthTokens.refresh_token,
		};
	}

	async refreshOAuthAccessToken(
		provider: TOAuthProvider,
		oAuthRefreshToken: string
	) {
		const oAuthTokens = await this.fetchOAuthTokens(
			provider,
			"refresh_token",
			oAuthRefreshToken
		);
		return {
			oAuthAccessToken: oAuthTokens.access_token,
		};
	}

	private async fetchOAuthTokens(
		provider: TOAuthProvider,
		grantType: TOAuthTokenRequestGrantType,
		grantValue: string
	) {
		const oAuthTokenResponse = await fetch(
			this.oAuthProps[provider].requestEndpoint.token,
			buildTokenFetchOptions(
				provider,
				grantType,
				grantValue,
				this.oAuthProps
			)
		);

		if (oAuthTokenResponse.status >= 500) {
			throw ServerError.etcError(
				500,
				OAUTH_TOKEN_SERVICE_ERROR_MESSAGES.OAUTH_PROVIDER_ERROR
			);
		} else if (oAuthTokenResponse.status >= 400) {
			throw ServerError.badRequest(
				OAUTH_TOKEN_SERVICE_ERROR_MESSAGES.OAUTH_PROVIDER_INVALID
			);
		}

		return (await oAuthTokenResponse.json()) as IOAuthTokens;
	}

	private async fetchOAuthUserByAccessToken(
		provider: TOAuthProvider,
		accessToken: string
	) {
		const url = this.oAuthProps[provider].requestEndpoint.user;
		const headers = {
			Authorization: `Bearer ${accessToken}`,
			"Content-type": oAuthRequestContentType,
		};

		const oAuthUserResponse = await fetch(url, { headers });
		return await oAuthUserResponse.json();
	}
}
