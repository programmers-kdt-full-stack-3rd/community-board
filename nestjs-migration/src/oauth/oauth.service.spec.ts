import { Test, TestingModule } from "@nestjs/testing";
import { TOAuthProvider } from "shared";
import { UpdateResult } from "typeorm";
import { AuthService } from "../auth/auth.service";
import { RefreshTokensRepository } from "../auth/refresh-tokens.repository";
import { ServerError } from "../common/exceptions/server-error.exception";
import { User } from "../user/entities/user.entity";
import { UserRepository } from "../user/user.repository";
import * as oAuthUtil from "../utils/oauth.util";
import { OAuthPropsConfig } from "./config/oauth-props.config";
import { OAuthLoginDto } from "./dto/oauth-login.dto";
import { OAuthConnection } from "./entities/oauth-connection.entity";
import { OAuthProvider } from "./entities/oauth-provider.entity";
import { TOAuthLoginType, TOAuthProps } from "./interfaces/oauth.interface";
import { OAuthService } from "./oauth.service";
import { OAuthTokenService } from "./oauthtoken.service";
import { OAuthConnectionRepository } from "./repositories/oauth-connection.repository";
import { OAuthProviderRepository } from "./repositories/oauth-provider.repository";

jest.mock("typeorm-transactional", () => ({
	Transactional: () => () => ({}),
}));

describe("OAuthService", () => {
	let oAuthService: OAuthService;
	let oAuthPropsConfig: OAuthPropsConfig;
	let authService: AuthService;
	let userRepository: UserRepository;
	let oAuthProviderRepository: OAuthProviderRepository;
	let oAuthConnectionRepository: OAuthConnectionRepository;
	let refreshTokenRepository: RefreshTokensRepository;
	let oAuthTokenService: OAuthTokenService;

	const mockProvider: TOAuthProvider = "google";
	const mockAuthProps: Partial<Record<TOAuthProvider, TOAuthProps>> = {
		google: {
			clientId: "googleClientId",
			clientSecret: "googleClientSecret",
			scope: "https://www.googleapis.com/auth/userinfo.profile",
			redirectUri: "http://localhost:3000/oauth/redirect/google",
			requestEndpoint: {
				login: "https://accounts.google.com/o/oauth2/v2/auth",
				token: "https://oauth2.googleapis.com/token",
				user: "https://www.googleapis.com/oauth2/v2/userinfo",
				revoke: "https://oauth2.googleapis.com/revoke",
			},
			getAdditionalRequestOptionsFor: {
				login: () => ({
					searchParams: {
						access_type: "offline",
					},
				}),

				revoke: () => ({
					searchParams: {
						access_type: "offline",
					},
				}),
			},
			reconfirmParams: { prompt: "consent" },
		},
	};

	const createMockUser = (overrides: Partial<User> = {}): User => {
		const user = new User();
		return Object.assign(
			user,
			{
				id: 1,
				nickname: "testuser",
			} as User,
			overrides
		);
	};

	beforeEach(async () => {
		const module: TestingModule = await Test.createTestingModule({
			providers: [
				OAuthService,
				{
					provide: OAuthPropsConfig,
					useValue: {
						getOAuthProps: jest.fn().mockReturnValue(mockAuthProps),
					},
				},
				{
					provide: AuthService,
					useValue: {
						generateTokens: jest.fn(),
					},
				},
				{
					provide: UserRepository,
					useValue: {
						readUserByOAuth: jest.fn(),
						save: jest.fn(),
					},
				},
				{
					provide: OAuthProviderRepository,
					useValue: {
						findOne: jest.fn(),
					},
				},
				{
					provide: OAuthConnectionRepository,
					useValue: {
						getOAuthConnectionByProviderAndAccountId: jest.fn(),
						update: jest.fn(),
						insert: jest.fn(),
					},
				},
				{
					provide: RefreshTokensRepository,
					useValue: {
						save: jest.fn(),
					},
				},
				{
					provide: OAuthTokenService,
					useValue: {
						verifyAuthorizationCode: jest.fn(),
					},
				},
			],
		}).compile();

		oAuthService = module.get<OAuthService>(OAuthService);
		oAuthPropsConfig = module.get<OAuthPropsConfig>(OAuthPropsConfig);
		authService = module.get<AuthService>(AuthService);
		userRepository = module.get<UserRepository>(UserRepository);
		oAuthProviderRepository = module.get<OAuthProviderRepository>(
			OAuthProviderRepository
		);
		oAuthConnectionRepository = module.get<OAuthConnectionRepository>(
			OAuthConnectionRepository
		);
		refreshTokenRepository = module.get<RefreshTokensRepository>(
			RefreshTokensRepository
		);
		oAuthTokenService = module.get<OAuthTokenService>(OAuthTokenService);
	});

	it("should be defined", () => {
		expect(oAuthService).toBeDefined();
	});

	describe("getOAuthUrl", () => {
		const verifyUrlParams = (
			url: string,
			expectedParams: Record<string, string>
		) => {
			const parsedUrl = new URL(url);
			Object.entries(expectedParams).forEach(([key, value]) => {
				expect(parsedUrl.searchParams.get(key)).toBe(value);
			});
		};

		it.each([
			["login", "login_type%3Dlogin"],
			["reconfirm", "login_type%3Dreconfirm"],
			["link", "login_type%3Dlink"],
		])(
			"%s 타입에 대해 올바른 OAuth URL을 반환한다",
			(loginType, expectedState) => {
				const url = oAuthService.getOAuthUrl(
					loginType as TOAuthLoginType,
					"google"
				);

				verifyUrlParams(url, {
					response_type: "code",
					client_id: "googleClientId",
					redirect_uri: "http://localhost:3000/oauth/redirect/google",
					scope: "https://www.googleapis.com/auth/userinfo.profile",
					access_type: "offline",
				});

				expect(url).toContain(`state=${expectedState}`);

				if (loginType === "reconfirm") {
					expect(url).toContain("prompt=consent");
				}
			}
		);

		it("생성된 URL이 예상 구조와 일치한다", () => {
			const loginUrl = oAuthService.getOAuthUrl("login", "google");
			expect(loginUrl).toMatchSnapshot();
		});

		it("getAdditionalRequestOptionsFor가 없는 경우에도 정상적으로 동작한다", () => {
			const modifiedMockProps = { ...mockAuthProps };
			delete modifiedMockProps.google.getAdditionalRequestOptionsFor
				.login;

			jest.spyOn(oAuthPropsConfig, "getOAuthProps").mockReturnValue(
				modifiedMockProps as Record<TOAuthProvider, TOAuthProps>
			);

			const url = oAuthService.getOAuthUrl("login", "google");

			const parsedUrl = new URL(url);

			verifyUrlParams(url, {
				response_type: "code",
				client_id: "googleClientId",
				redirect_uri: "http://localhost:3000/oauth/redirect/google",
				scope: "https://www.googleapis.com/auth/userinfo.profile",
			});

			expect(parsedUrl.searchParams.get("access_type")).toBeNull();
		});

		it("searchParams가 없는 경우에도 정상적으로 동작한다", () => {
			const modifiedMockProps = { ...mockAuthProps };
			modifiedMockProps.google.getAdditionalRequestOptionsFor = {
				login: () => ({}),
			};

			jest.spyOn(oAuthPropsConfig, "getOAuthProps").mockReturnValue(
				modifiedMockProps as Record<TOAuthProvider, TOAuthProps>
			);

			const url = oAuthService.getOAuthUrl("reconfirm", "google");

			const parsedUrl = new URL(url);

			verifyUrlParams(url, {
				response_type: "code",
				client_id: "googleClientId",
				redirect_uri: "http://localhost:3000/oauth/redirect/google",
				scope: "https://www.googleapis.com/auth/userinfo.profile",
			});

			expect(parsedUrl.searchParams.get("access_type")).toBeNull();
		});
	});

	describe("oAuthLogin", () => {
		const mockTokens = {
			accessToken: "access-token",
			refreshToken: "refresh-token",
		};

		const moackOAuthAccountId = "oauth-account-id";
		const mockOAuthRefreshToken = "oauth-refresh-token";

		const mockOAuthData = {
			oAuthAccountId: moackOAuthAccountId,
			oAuthRefreshToken: mockOAuthRefreshToken,
		};

		beforeEach(() => {
			jest.spyOn(
				oAuthTokenService,
				"verifyAuthorizationCode"
			).mockResolvedValue(mockOAuthData);

			jest.spyOn(authService, "generateTokens").mockReturnValue(
				mockTokens
			);
		});

		const oAuthLoginDto: OAuthLoginDto = {
			provider: "google",
			code: "code",
		};

		describe("성공 케이스", () => {
			it("기존 사용자 로그인 성공", async () => {
				const mockUser = createMockUser();

				jest.spyOn(userRepository, "readUserByOAuth").mockResolvedValue(
					mockUser
				);

				jest.spyOn(
					oAuthConnectionRepository,
					"getOAuthConnectionByProviderAndAccountId"
				).mockResolvedValue({ id: 1 } as OAuthConnection);

				jest.spyOn(
					oAuthConnectionRepository,
					"update"
				).mockResolvedValue({ affected: 1 } as UpdateResult);

				const result = oAuthService.oAuthLogin(oAuthLoginDto);

				await expect(result).resolves.toEqual({
					nickname: mockUser.nickname,
					accessToken: mockTokens.accessToken,
					refreshToken: mockTokens.refreshToken,
				});
			});

			it("새로운 사용자 로그인 성공", async () => {
				jest.spyOn(userRepository, "readUserByOAuth").mockResolvedValue(
					null
				);

				jest.spyOn(oAuthUtil, "generateNickname").mockReturnValue(
					"testuser"
				);

				jest.spyOn(userRepository, "save").mockResolvedValue(
					createMockUser({ nickname: "testuser" })
				);

				jest.spyOn(
					oAuthProviderRepository,
					"findOne"
				).mockResolvedValue({
					id: 1,
				} as OAuthProvider);

				const result = await oAuthService.oAuthLogin(oAuthLoginDto);

				expect(result).toEqual({
					nickname: "testuser",
					accessToken: mockTokens.accessToken,
					refreshToken: mockTokens.refreshToken,
				});
			});
		});

		describe("실패 케이스", () => {
			it("탈퇴한 회원인 경우 에러를 반환한다", async () => {
				const mockUser = createMockUser({ isDelete: true });
				jest.spyOn(userRepository, "readUserByOAuth").mockResolvedValue(
					mockUser
				);

				const result = oAuthService.oAuthLogin(oAuthLoginDto);

				await expect(result).rejects.toThrow(ServerError);
				await expect(result).rejects.toThrow("탈퇴한 회원입니다.");
			});

			it("연동된 소셜 계정이 없는 경우 에러를 반환한다", async () => {
				const mockUser = createMockUser();

				jest.spyOn(userRepository, "readUserByOAuth").mockResolvedValue(
					mockUser
				);

				jest.spyOn(
					oAuthConnectionRepository,
					"getOAuthConnectionByProviderAndAccountId"
				).mockResolvedValue(null);

				const result = oAuthService.oAuthLogin(oAuthLoginDto);

				await expect(result).rejects.toThrow(ServerError);
				await expect(result).rejects.toThrow(
					"연동된 소셜 계정이 없습니다."
				);
			});

			it("Refresh token 저장에 실패한 경우 에러를 반환한다", async () => {
				const mockUser = createMockUser();

				jest.spyOn(userRepository, "readUserByOAuth").mockResolvedValue(
					mockUser
				);

				jest.spyOn(
					oAuthConnectionRepository,
					"getOAuthConnectionByProviderAndAccountId"
				).mockResolvedValue({ id: 1 } as OAuthConnection);

				jest.spyOn(
					oAuthConnectionRepository,
					"update"
				).mockResolvedValue({ affected: 0 } as UpdateResult);

				const result = oAuthService.oAuthLogin(oAuthLoginDto);

				await expect(result).rejects.toThrow(ServerError);
				await expect(result).rejects.toThrow(
					"OAuth Refresh token 저장 실패"
				);
			});

			it("이미 연동된 소셜 계정인 경우 에러를 반환한다", async () => {
				const mockUser = createMockUser();

				jest.spyOn(userRepository, "readUserByOAuth").mockResolvedValue(
					mockUser
				);

				jest.spyOn(
					oAuthConnectionRepository,
					"getOAuthConnectionByProviderAndAccountId"
				).mockResolvedValue({ id: 1 } as OAuthConnection);

				jest.spyOn(
					oAuthConnectionRepository,
					"update"
				).mockRejectedValue({ code: "ER_DUP_ENTRY" });

				const result = oAuthService.oAuthLogin(oAuthLoginDto);

				await expect(result).rejects.toThrow(ServerError);
				await expect(result).rejects.toThrow(
					"이미 연동된 소셜 계정입니다."
				);
			});

			it("새로운 소셜로그인 유저를 저장 실패 할 경우 에러를 반환한다.", async () => {
				jest.spyOn(userRepository, "readUserByOAuth").mockResolvedValue(
					null
				);

				jest.spyOn(userRepository, "save").mockResolvedValue(null);

				const result = oAuthService.oAuthLogin(oAuthLoginDto);

				await expect(result).rejects.toThrow(ServerError);
				await expect(result).rejects.toThrow(
					"소셜 로그인으로 회원가입 실패"
				);
			});

			it("db에 없는 프로바이더인경우 오류를 반환한다.", async () => {
				jest.spyOn(userRepository, "readUserByOAuth").mockResolvedValue(
					null
				);

				jest.spyOn(userRepository, "save").mockResolvedValue(
					createMockUser()
				);

				jest.spyOn(
					oAuthProviderRepository,
					"findOne"
				).mockResolvedValue(null);

				const result = oAuthService.oAuthLogin(oAuthLoginDto);

				await expect(result).rejects.toThrow(ServerError);
				await expect(result).rejects.toThrow(
					"소셜 로그인 연동에 실패했습니다."
				);
			});

			it("소셜계정으로 회원가입 중 db에 연동된 소셜계정이 있는 경우 에러를 반환한다.", async () => {
				jest.spyOn(userRepository, "readUserByOAuth").mockResolvedValue(
					null
				);

				jest.spyOn(oAuthUtil, "generateNickname").mockReturnValue(
					"testuser"
				);

				jest.spyOn(userRepository, "save").mockResolvedValue(
					createMockUser({ nickname: "testuser" })
				);

				jest.spyOn(
					oAuthProviderRepository,
					"findOne"
				).mockRejectedValue({ code: "ER_DUP_ENTRY" });

				const result = oAuthService.oAuthLogin(oAuthLoginDto);

				await expect(result).rejects.toThrow(ServerError);
				await expect(result).rejects.toThrow(
					"이미 연동된 소셜 계정입니다."
				);
			});

			it("소셜계정으로 회원가입 중 중복된 닉네임이 있는 경우 에러를 반환한다.", async () => {
				jest.spyOn(userRepository, "readUserByOAuth").mockResolvedValue(
					null
				);

				jest.spyOn(userRepository, "save").mockRejectedValue({
					code: "ER_DUP_ENTRY",
				});

				const result = oAuthService.oAuthLogin(oAuthLoginDto);

				await expect(result).rejects.toThrow(ServerError);
				await expect(result).rejects.toThrow(
					"이미 사용 중인 닉네임입니다."
				);
			});

			it("로그인시 user.id가 없는 경우 에러를 반환한다.", async () => {
				const mockUser = createMockUser({ id: null });

				jest.spyOn(userRepository, "readUserByOAuth").mockResolvedValue(
					mockUser
				);

				jest.spyOn(
					oAuthConnectionRepository,
					"getOAuthConnectionByProviderAndAccountId"
				).mockResolvedValue({ id: 1 } as OAuthConnection);

				jest.spyOn(
					oAuthConnectionRepository,
					"update"
				).mockResolvedValue({ affected: 1 } as UpdateResult);

				const result = oAuthService.oAuthLogin(oAuthLoginDto);

				await expect(result).rejects.toThrow(ServerError);
				await expect(result).rejects.toThrow("사용자 정보 오류");
			});

			it("로그인시 user.nickname이 없는 경우 에러를 반환한다.", async () => {
				const mockUser = createMockUser({ nickname: null });

				jest.spyOn(userRepository, "readUserByOAuth").mockResolvedValue(
					mockUser
				);

				jest.spyOn(
					oAuthConnectionRepository,
					"getOAuthConnectionByProviderAndAccountId"
				).mockResolvedValue({ id: 1 } as OAuthConnection);

				jest.spyOn(
					oAuthConnectionRepository,
					"update"
				).mockResolvedValue({ affected: 1 } as UpdateResult);

				const result = oAuthService.oAuthLogin(oAuthLoginDto);

				await expect(result).rejects.toThrow(ServerError);
				await expect(result).rejects.toThrow("사용자 정보 오류");
			});
		});
	});
});
