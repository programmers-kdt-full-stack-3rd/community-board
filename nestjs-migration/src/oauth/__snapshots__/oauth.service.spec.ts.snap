// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OAuthService getOAuthUrl 생성된 URL이 예상 구조와 일치한다 1`] = `"https://accounts.google.com/o/oauth2/v2/auth?response_type=code&client_id=googleClientId&redirect_uri=http%3A%2F%2Flocalhost%3A3000%2Foauth%2Fredirect%2Fgoogle&state=login_type%3Dlogin&scope=https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fuserinfo.profile&access_type=offline"`;
